File: service/research/visualizations/registry.py
class VisualizationRegistry:
    """Registry that lets new visualization commands be added without branching."""

    def __init__(self, commands=None):
        self._commands = {}
        for command in commands or []:
            self.register(command)

    def register(self, command):
        if command.name in self._commands:
            raise ValueError(f"Duplicate visualization command: {command.name}")
        self._commands[command.name] = command

    def all(self):
        return list(self._commands.values())

File: service/src/research/visualizations/registry.ts
import type { VisualizationCommand } from "./runner.js";

export class VisualizationRegistry {
  private readonly commands = new Map<string, VisualizationCommand>();

  constructor(commands: VisualizationCommand[] = []) {
    commands.forEach((command) => this.register(command));
  }

  register(command: VisualizationCommand): void {
    if (this.commands.has(command.name)) throw new Error(`Duplicate visualization command: ${command.name}`);
    this.commands.set(command.name, command);
  }

  all(): VisualizationCommand[] { return [...this.commands.values()]; }
}
